enhancements/ingress/dynamic-config-manager.md

```diff
@@ -0,0 +1,469 @@
+
+Answer: **Probably yes**.
+
+Enabling Dynamic Config Manager should not cause any regressions.  We need to
+fix known issues regarding inaccurate metrics and session stickiness and any
+other regressions that we find
+(see [Risks and Mitigations](#risks-and-mitigations)).
+
+#### Do we need an override to turn off Dynamic Config Manager?
+
+Answer: **Probably yes**.
+
+Ingress is a critical cluster function, and it is an extremely
+performance-sensitive one for some customers.  In case we ship DCM enabled and a
+customer finds some critical issue, we need an unsupported config override that
+support staff can use to turn off DCM temporarily until we fix the issue.
+
+### Test Plan
+
+Dynamic Config Manager will be enabled initially using a TechPreviewNoUpgrade
+featuregate to provide CI signal.  We will additionally revisit existing E2E
+tests, enable any DCM-related E2E tests that are currently not enabled, and
+verify that all enabled router-related E2E tests pass with DCM.  As known
+defects are fixed and any unknown ones are discovered, we will add additional
+tests.  Finally, we will work with partners to verify DCM for their use-cases.
+
+Expanding test coverage is important both for verifying that DCM is ready for GA
+as well as for enabling us to improve DCM in future releases with confidence
+that we are not introducing regressions.
+
+### Graduation Criteria
+
+We will introduce the feature as TechPreviewNoUpgrade in OpenShift 4.18, with
+the goal of graduating it to GA in the same release.  Further improvements to
+DCM will follow in subsequent OpenShift releases.
+
+#### Dev Preview -> Tech Preview
+
+N/A.
+
+#### Tech Preview -> GA
+
+- All known regressions are fixed.
+- Payload tests pass with DCM enabled.
+- At least 2 partners or customers provide test results on DCM's function.
+- Memory overhead is within acceptable limits.
+- Limitations (such as which changes still require reload) are documented.
+
+#### Removing a deprecated feature
+
+N/A.
+
+### Upgrade / Downgrade Strategy
+
+The feature requires no specific considerations for upgrade or downgrade;
+cluster-ingress-operator will handle upgrading the router image and configuring
+Dynamic Config Manager, using the standard process for a rolling update of the
+router deployment.
+
+### Version Skew Strategy
+
+N/A.
+
+### Operational Aspects of API Extensions
+
+N/A.  The operation of this feature should be transparent to the end user.
+
+#### Failure Modes
+
+If Dynamic Config Manager cannot handle a configuration change, it falls back to
+the old fork-and-reload procedure.  Thus in the worst case, OpenShift router
+should behave no better or worse with DCM than without DCM.
+
+If an unforeseen defect arises, DCM can be inhibited using an unsupported config
+override:
+
+```shell
+oc -n openshift-ingress-operator patch ingresscontrollers/default --type=merge --patch='{"spec":{"unsupportedConfigOverrides":{"dynamicConfigManager":"false"}}}'
```

Comment:
For upgrades, I assume you need to remove the unsupported override and re-enable it once the upgrade is complete. However, this raises the question: if DCM is re-enabled, does that imply the cluster may be broken from an ingress perspective until the update is complete and DCM can be disabled again?  Is that going to be an acceptable stance for production clusters?

Reply:
> For upgrades, I assume you need to remove the unsupported override and re-enable it once the upgrade is complete. 

We *should*, but currently do not, block upgrades when using `unsupportedConfigOverrides` (reported as https://issues.redhat.com/browse/OCPBUGS-41366).  

> However, this raises the question: if DCM is re-enabled, does that imply the cluster may be broken from an ingress perspective until the update is complete and DCM can be disabled again? Is that going to be an acceptable stance for production clusters?

I guess you're saying that `unsupportedConfigOverrides` isn't a good escape hatch after all, right?  You can probably force an upgrade using `oc adm --force`, but I understand that `--force` is a last resort.  

I suppose we could tell people to turn off the featuregate.  Is that a reasonable escape hatch?

Reply:
> I suppose we could tell people to turn off the featuregate. Is that a reasonable escape hatch?

Yes, I would say so.